wait_for_flow_run PrefectClient function

Fizzizist opened the pull request:
<!-- 
Thanks for opening a pull request to Prefect! We've got a few requests to help us review contributions:

- Make sure that your title neatly summarizes the proposed changes.
- Provide a short overview of the change and the value it adds.
- Share an example to help us understand the change in user experience.
- Confirm that you've done common tasks so we can give a timely review.
- Review our contribution guidelines: https://docs.prefect.io/latest/contributing/overview/

Happy engineering!
-->

As described in issue #11177 , I would like to see this feature in production. The ability to use the PrefectClient to automatically wait for a flow run to be done before proceeding, using just the prefect client.

closes #11177 

### Example
Imagine you have an API endpoint that fires off a flow_run and then creates a background task that waits for the flow run to finish so it can use the results.
```python
async def my_endpoint(background_tasks, flow_input):
    async with get_client() as c:
        new_flow_run = await c.create_flow_run_from_deployment('<some UUID>', flow_input)
    background_tasks.append(wait_and_insert, new_flow_run)

async def wait_and_insert(flow_run):
    async with get_client() as c:
        completed_flow_run = await c.wait_for_flow_run(flow_run.id)
    if completed_flow_run.state.type = StateType.COMPLETED:
        result = completed_flow_run.state.result(fetch=True)
        insert_results(result)
    else:
        raise Failure
````
Suppose here that `insert_results` is some database write function.

### Checklist
<!-- These boxes may be checked after opening the pull request. -->

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.
  <!-- If you do not have permission to add a label, a maintainer will add one for you -->

For documentation changes:

- [ ] This pull request includes redirect settings in `netlify.toml` for files that are removed or renamed

desertaxle commented:
Thanks for the contribution @Fizzizist! Our client methods are generally thin wrappers around API calls without extra logic, so I think this functionality should live in a different module. I think it makes sense to move this to a submodule under `prefect.utilities`, but I'd like to get a second opinion from @serinamarie.

desertaxle commented:
Hey @Fizzizist, @serinamarie and I discussed and we think it'd be best to move this new functionality into a `prefect.flow_runs` module. Let us know if there's any way we can help with this PR!

Fizzizist commented:
> Hey @Fizzizist, @serinamarie and I discussed and we think it'd be best to move this new functionality into a `prefect.flow_runs` module. Let us know if there's any way we can help with this PR!

@desertaxle sounds good. So, just to be clear. You want a new module at `src/prefect/flow_runs.py` to hold this functionality?

desertaxle commented:
> You want a new module at `src/prefect/flow_runs.py` to hold this functionality?

That's correct!

Fizzizist commented:
@desertaxle i have pushed a change and moved the function. I also moved the tests. I am having an issue running the tests now for a seemingly unrelated reason. I am getting a DeprecationWarning about the `event_loop` fixture that is defined in the conftest.py module, and it is making my test fail. Do you have any insights about this? Thanks

<img width="805" alt="image" src="https://github.com/PrefectHQ/prefect/assets/16968180/e9b9796e-093c-4023-a85f-64dd66f3e7f0">

desertaxle commented:
There was a recent update to `pytest-asyncio` that doesn't work well with some of our fixtures. If you install `pytest-asyncio<0.23.0` then that warning should go away.

Fizzizist commented:
@desertaxle thanks! tests are passing on my end. Changes are ready for review. Let me know if there is anything else you'd like me to change/add :)

desertaxle commented on src/prefect/flow_runs.py:
```diff
@@ -0,0 +1,37 @@
+async def wait_for_flow_run(
+    client: PrefectClient,
+    flow_run_id: UUID,
+    timeout: int = 10800,
```
It'd be nice to type this a `Optional` since passing `None` will disable the timeout.
```suggestion
    timeout: Optional[int] = 10800,
```

Fizzizist replied:
addressed

Fizzizist commented on src/prefect/flow_runs.py:
```diff
@@ -0,0 +1,37 @@
+            await anyio.sleep(poll_interval)
+
+    raise FlowWaitTimeout(
+        f"FlowRun with ID {flow_run_id} is taking longer than {timeout} seconds"
```
addressed

Fizzizist commented on src/prefect/exceptions.py:
```diff
@@ -401,3 +401,7 @@ class NotPausedError(PrefectException):
     """Raised when a flow pause times out"""
+
+
+class FlowWaitTimeout(PrefectException):
```
addressed

desertaxle reviewed (approved):
LGTM!

serinamarie commented:
@desertaxle Do we want any docs to go out for this, or are we good with the docstring? I think if latter, maybe we should add the example to the `wait_for_flow_run` docstring?

serinamarie commented:
(also happy to take on either)

desertaxle commented:
@serinamarie I'll open a PR to add some docs
